microsoft/PowerToys
# [Auto-update] Auto-update improvements

## Summary of the Pull Request

**What is this about:**

* Refactored auto-update mechanism to use Settings window buttons.
* State is stored in the `AppData\Local\Microsoft\PowerToys\UpdateState.json`. 
* Updated UI in the Settings
* Added file watcher to track changes in the `UpdateState` file.

PT Settings UI for different states (updated):

![0](https://user-images.githubusercontent.com/8949536/118967724-f8616c80-b962-11eb-82bb-920fcd82bcf7.PNG)

_After the `Check for updates` button is pressed and there are no updates:_

![image](https://user-images.githubusercontent.com/3206696/119102921-d9beac80-ba1a-11eb-9ac1-f4fc9d28bbc1.png)

![2](https://user-images.githubusercontent.com/8949536/118967736-fac3c680-b962-11eb-8e3b-fec0ea761d10.PNG)
![3](https://user-images.githubusercontent.com/8949536/118967738-fb5c5d00-b962-11eb-9b8c-c9d1871e8328.PNG)

_If an error occurred (including the case the user cancelled the installation):_

![image](https://user-images.githubusercontent.com/3206696/119105192-5c486b80-ba1d-11eb-9cb8-fa458503dba3.png)


**What is include in the PR:** 

**How does someone test / validate:** 

* Update `src/Version.props`, set any version below 0.37.2 for testing (e.g. <Version>0.35.1</Version>)
* Clean build PT Release
* Build `installer/PowerToysSetup`
* Install `installer/PowerToysSetup/x64/Release/PowerToysSetup-0.35.1-x64.msi`
* Verify behavior after clicking the buttons in different states with `Download updates automatically` switched on and off. 

## Quality Checklist

- [x] **Linked issue:** #10589

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.
